subsys/usb/device/class/cdc_acm.c

```diff
@@ -256,6 +256,7 @@ static void tx_work_handler(struct k_work *work)
 	uint8_t ep = cfg->endpoint[ACM_IN_EP_IDX].ep_addr;
```

Comment:
As I say in the initial commit (which I was forced to split and shorten):

usb_transfer() drops data by not checking the return value of k_work_submit_to_queue()
This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again. 

The stuck after reset comes from:
cdc_acm_irq_tx_ready() did not check configured and suspended states 
-> This caused the device to show as ready even though it was not configured after a host reset
The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

Together these cause the issues described in the issue post. 
My changes changed this.

Reply:
> As I say in the initial commit (which I was forced to split and shorten):
> 
> usb_transfer() drops data by not checking the return value of k_work_submit_to_queue() This causes tx_work_handler() to not send the data because it thinks they are in the work queue now. It also thinks the device is occupied but as there is no work queue item it won't get freed again.

Can you please provide k_work_submit_to_queue() return value instead of your own interpretation?

> The stuck after reset comes from: cdc_acm_irq_tx_ready() did not check configured and suspended states -> This caused the device to show as ready even though it was not configured after a host reset The same can happen for RX so I fixed it with it, it's the same issue of ignored state.

But it is okay to fill/read the UART FIFO (ringbuffer) event if USB device is not configured or suspended.

> 
> Together these cause the issues described in the issue post. My changes changed this.

I can help you write a commit message later, after we figure out what caused the issue.

Reply:
> I can help you write a commit message later, after we figure out what caused the issue.

That's great, learning to fit better into the zephyr workflow will be helpful.  :)
It may delay because today is my last day before the holidays but with some luck it may be done before